jbmc/unit/java_bytecode/java_bytecode_instrument/virtual_call_null_checks.cpp

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************\
+
+ Module: Unit test to check Java virtual calls via a pointer
+         yield a correct sequence of not-null assumptions.
+
+ Author: Diffblue Limited.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <testing-utils/message.h>
+#include <java-testing-utils/load_java_class.h>
+
+#include <analyses/local_safe_pointers.h>
+#include <goto-programs/goto_convert_functions.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <util/expr_iterator.h>
+
+SCENARIO(
+  "java_bytecode_virtual_call_null_checks",
+  "[core][java_bytecode][java_bytecode_instrument]")
+{
+  GIVEN("A class that makes a virtual call via a pointer")
+  {
+    symbol_tablet symbol_table = load_java_class(
+      "VirtualCallNullChecks", "./java_bytecode/java_bytecode_instrument");
+    namespacet ns(symbol_table);
+    goto_functionst goto_functions;
+    goto_convert(symbol_table, goto_functions, null_message_handler);
+
+    WHEN("The virtual call is converted")
+    {
+      const auto &main_function =
+        goto_functions.function_map.at("java::VirtualCallNullChecks.main:()V");
+
+      THEN("The callee should be of the form some_ptr->field")
+      {
```

Comment:
:question: I'm not sure on the point of this `THEN` block - is it just to check that the first function call is what you're expecting? :pick: It might be a little clearer to use `std::find_if` and perhaps a check that is really is the function call you're looking for (i.e. the identifier is `java::B.virtualmethod:()V`)

Reply:
Yes, it's to stop the test passing because I loaded an empty class file or the compiler optimised the call away or similar. I'll add a comment.